BezierKit/Library/AugmentedGraph.swift

```diff
@@ -282,27 +308,48 @@ private extension AugmentedGraph {
         }
         nodes = Array(nodes[0...currentUniqueIndex])
     }
-    func findUnvisitedPath(from node: Node, to goal: Node) -> [(Edge, Bool)]? {
+    func findUnvisitedPath(from node: Node, to goal: Node, preferring: Edge.Solution) -> [(Edge, Bool)]? {
+        let preferredDirection: Edge.Solution = (preferring == .clockwise) ? .clockwise : .counterClockwise
+        let nonPreferredDirection: Edge.Solution = (preferring == .clockwise) ? .counterClockwise : .clockwise
         func pathUsingEdge(_ edge: Edge?, from node: Node, forwards: Bool) -> [(Edge, Bool)]? {
             guard let edge = edge, edge.needsVisiting else { return nil }
             edge.visited = true
             edge.visitCoincidentEdges()
             let nextNode = forwards ? edge.endingNode : edge.startingNode
-            if let path = findUnvisitedPath(from: nextNode, to: goal) {
+            // we know which direction we went and our preferred winding direction, we can keep the same way
+            let nextPreferring = forwards ? preferredDirection : nonPreferredDirection
+            if let path = findUnvisitedPath(from: nextNode, to: goal, preferring: nextPreferring) {
                 return [(edge, forwards)] + path
             } else {
                 return nil
             }
         }
-        // we prefer to keep the direction of the path the same which is why
-        // we try all the possible forward edges before any back edges
-        if let result = pathUsingEdge(node.forwardEdge, from: node, forwards: true) { return result }
-        for neighbor in node.neighbors {
-            if let result = pathUsingEdge(neighbor.forwardEdge, from: neighbor, forwards: true) { return result }
+        func priority(_ tuple: (edge: Edge, forwards: Bool)) -> Int {
+            let edge = tuple.edge
+            let forwards = tuple.forwards
+            if forwards, edge.inSolution == preferredDirection {
+                return 3
+            } else if !forwards, edge.inSolution == nonPreferredDirection {
+                return 2
+            } else if forwards {
+                return 1
+            } else {
+                return 0
+            }
         }
-        if let result = pathUsingEdge(node.backwardEdge, from: node, forwards: false) { return result }
-        for neighbor in node.neighbors {
-            if let result = pathUsingEdge(neighbor.backwardEdge, from: neighbor, forwards: false) { return result }
+        // go through all unvisited edges in the solution
+        // if we prefer clockwise take any foward edges with .clockwise, or backwards edges with .counterClockwise
+        // if we prefer counterClockwise take any forward edges with .counterClockwise, or backwards edges with .clockWise
+        var candidateList = node.allIncidentEdges
+        candidateList.sort { priority($1) > priority($0) }
```

Comment:
instead of sorting we could just find the maximum priority unvisited edge. Once we call `pathUsingEdge` the edge is marked visited, so if it doesn't work, just call it again.